File: glasgow/decks/code/equivalence/equivalence.R
################################################################################
# name:        equivalence.R
# author:      scott cunningham
# description: Five equivalent paths to the 2x2 DiD on castle.dta.
#              Manual 2x2, OLS with interactions, TWFE, first-difference on
#              treatment dummy, and across-group difference on post dummy.
#              All five return the same DiD coefficient.
################################################################################

# install.packages(c("tidyverse", "fixest", "haven"))
library(tidyverse)
library(fixest)
library(haven)

# Load and filter the castle data: keep states adopting in 2006 + never-treated
data <- haven::read_dta(
  "https://github.com/scunning1975/mixtape/raw/master/castle.dta"
) %>%
  filter(!(effyear %in% c(2005, 2007, 2008, 2009))) %>%
  mutate(
    year  = as.numeric(year),
    post  = ifelse(year >= 2006, 1, 0),
    treat = ifelse(!is.na(effyear), 1, 0)
  ) %>%
  filter(year %in% c(2005, 2006))


# ----------------------------------------------------------------------------
# PATH 1: Manual 2x2 (Snow's "long differences": four averages, three subs)
# ----------------------------------------------------------------------------
cells <- data %>%
  group_by(treat, post) %>%
  summarise(y = mean(l_homicide, na.rm = TRUE), .groups = "drop")
print(cells)

did_manual <- with(cells, {
  (y[treat == 1 & post == 1] - y[treat == 1 & post == 0]) -
  (y[treat == 0 & post == 1] - y[treat == 0 & post == 0])
})
cat("Path 1 (Manual 2x2):", did_manual, "\n")


# ----------------------------------------------------------------------------
# PATH 2: OLS with treat x post interaction ("saturated" spec)
# ----------------------------------------------------------------------------
path2 <- feols(
  l_homicide ~ post + treat + post:treat,
  data = data, cluster = ~ sid
)
print(path2)


# ----------------------------------------------------------------------------
# PATH 3: TWFE — state and year fixed effects, DD on treat:post
# ----------------------------------------------------------------------------
path3 <- feols(
  l_homicide ~ treat:post | sid + year,
  data = data, cluster = ~ sid
)
print(path3)


# ----------------------------------------------------------------------------
# PATH 4: First-difference within unit (Delta Y_i) regressed on treatment dummy
# ----------------------------------------------------------------------------
fd <- data %>%
  select(sid, year, treat, l_homicide) %>%
  pivot_wider(names_from = year, values_from = l_homicide,
              names_prefix = "y") %>%
  mutate(diff = y2006 - y2005)

path4 <- feols(diff ~ treat, data = fd, cluster = ~ sid)
print(path4)


# ----------------------------------------------------------------------------
# PATH 5: Across-group difference (treated - control by year) on post dummy.
#         FWL mirror of Path 4 — same coefficient.
# ----------------------------------------------------------------------------
across <- data %>%
  group_by(year, treat, post) %>%
  summarise(y = mean(l_homicide, na.rm = TRUE), .groups = "drop") %>%
  pivot_wider(names_from = treat, values_from = y, names_prefix = "treat_") %>%
  mutate(gap = treat_1 - treat_0)
print(across)

path5 <- lm(gap ~ post, data = across)
print(summary(path5))


# ----------------------------------------------------------------------------
# Population-weighted versions (sanity check that weights propagate)
# ----------------------------------------------------------------------------
path2_w <- feols(
  l_homicide ~ post + treat + post:treat,
  data = data, cluster = ~ sid, weights = ~ popwt
)
print(path2_w)

File: milano/Slides/code/equivalence.py
## equivalence.py
## ----------------------------------------------------------------------
## Three regression specifications for the 2x2 DiD on castle.dta,
## restricted to 2005 and 2006. All three return the same DiD coefficient.
##
## Mirrors equivalence.do (Stata) and equivalence.R (R).
## Run:  python3 equivalence.py
## ----------------------------------------------------------------------

import pandas as pd
import statsmodels.formula.api as smf

URL = "https://github.com/scunning1975/mixtape/raw/master/castle.dta"
df = pd.read_stata(URL)
df = df.query("year >= 2005 & year <= 2006").copy()
df["post"]  = (df.year == 2006).astype(int)
df["treat"] = (df.effyear == 2006).astype(int)

# ---- The four cell means ----
cells = df.groupby(["treat", "post"])["l_homicide"].mean().unstack()
y11 = cells.loc[1, 1]; y10 = cells.loc[1, 0]
y01 = cells.loc[0, 1]; y00 = cells.loc[0, 0]
did_by_hand = (y11 - y10) - (y01 - y00)
print(f"\nFour averages:\n  Y11={y11:.4f}  Y10={y10:.4f}  Y01={y01:.4f}  Y00={y00:.4f}")
print(f"  DiD by hand = {did_by_hand:.4f}\n")

# ---- Regression 1: OLS with interactions ----
m1 = smf.ols("l_homicide ~ post * treat", data=df).fit(
    cov_type="cluster", cov_kwds={"groups": df["sid"]})
print(f"Reg 1 (OLS interactions):  DiD = {m1.params['post:treat']:.4f}")

# ---- Regression 2: TWFE (state + year FE) ----
m2 = smf.ols("l_homicide ~ post:treat + C(sid) + C(year)", data=df).fit(
    cov_type="cluster", cov_kwds={"groups": df["sid"]})
print(f"Reg 2 (TWFE):              DiD = {m2.params['post:treat']:.4f}")

# ---- Regression 3: Long differences ----
wide = df.pivot(index="sid", columns="year", values="l_homicide").reset_index()
wide.columns = ["sid", "y2005", "y2006"]
treat_state = df.groupby("sid")["treat"].first().reset_index()
wide = wide.merge(treat_state, on="sid")
wide["diff"] = wide["y2006"] - wide["y2005"]
m3 = smf.ols("diff ~ treat", data=wide).fit(cov_type="HC1")
print(f"Reg 3 (Long differences):  DiD = {m3.params['treat']:.4f}")

print("\nAll three return the same DiD coefficient (to machine epsilon).")
